src/main/java/com/fasterxml/jackson/datatype/jdk8/OptionalSerializer.java

```diff
@@ -315,6 +315,7 @@ public void acceptJsonFormatVisitor(JsonFormatVisitorWrapper visitor, JavaType t
             }
             _dynamicSerializers = _dynamicSerializers.newWith(type, ser);
         }
+        ser = (JsonSerializer<Object>)provider.handlePrimaryContextualization(ser, _property);
```

Comment:
Unfortunately no test suite exists (that I know of) that uses `Date` types (or `Optional`); would be good to have.
However, I think I found bit different call for accessing serializer that will properly handle contextualization -- and as per discussion need to apply it via other repo.
Test case is from here, and this was super helpful in solving this, and in the process a class of problems for 2.9, where `ReferenceTypeSerializer` base class has the problem and can now be fixed.